cmd/chroma/main.go

```diff
@@ -160,17 +160,66 @@ func main() {
 		configureHTMLFormatter(ctx)
 	}
 	if len(cli.Files) == 0 {
-		contents, err := ioutil.ReadAll(os.Stdin)
+		data, err := ioutil.ReadAll(os.Stdin)
 		ctx.FatalIfErrorf(err)
-		format(ctx, w, style, lex(ctx, cli.Filename, string(contents)))
+
+		contents := string(data)
+		lexer := selexer(cli.Filename, contents)
+		if lexer == nil {
+			lexer = lexers.Fallback
+		}
+		format(ctx, w, style, lex(ctx, lexer, contents))
 	} else {
 		for _, filename := range cli.Files {
-			contents, err := ioutil.ReadFile(filename)
-			ctx.FatalIfErrorf(err)
 			if cli.Check {
-				check(filename, lex(ctx, filename, string(contents)))
+				data, err := ioutil.ReadFile(filename)
+				ctx.FatalIfErrorf(err)
+				contents := string(data)
+				lexer := selexer(filename, contents)
+				if lexer == nil {
+					lexer = lexers.Fallback
+				}
+				check(filename, lex(ctx, lexer, contents))
 			} else {
-				format(ctx, w, style, lex(ctx, filename, string(contents)))
+				fi, err := os.Stat(filename)
+				ctx.FatalIfErrorf(err)
+
+				var contents string
+				var lexer chroma.Lexer
+				peekSize := 1024
+				if !cli.Fail || peekSize > int(fi.Size()) {
```

Comment:
Would you mind pulling each branch into a function with a comment explaining what the logic is for each?